README.md
# rcap-go

![Release](https://github.com/md-irohas/rcap-go/actions/workflows/release.yml/badge.svg)
![Build and Tests](https://github.com/md-irohas/rcap-go/actions/workflows/build.yml/badge.svg)
[![codecov](https://codecov.io/gh/md-irohas/rcap-go/branch/master/graph/badge.svg?token=QMU3RTHEBE)](https://codecov.io/gh/md-irohas/rcap-go)

The `rcap` is a simple packet capturing utility written in Go.

I just want a utility to capture traffic for honeypot monitoring.
But existing utilities such as tcpdump and tshark do not satisfy my needs, so I wrote the `rcap`.

The `rcap` has the following functions.

* Dumping packets to files as PCAP format.
* Rotating pcap files every specified interval with offset (even if no packets are captured).
* Flexible filename format (timezone-aware).
* Random sampling of packets.
* Configuration file.
* Logging.


## Installation

### Compiled Binaries

Compiled binaries are available (only for Linux x86_64) at GitHub release page.

See https://github.com/md-irohas/rcap-go/releases.


### Manual Compilation

#### Requirements

* Go compiler
* libpcap-dev


#### Complilation

```sh
$ go build
```


## Usage

```sh
$ ./rcap --help
Usage of ./rcap:
  -L string
        [deprecated] log file.
  -S    use system time as a time source of rotation (default: use packet-captured time).
  -T int
        rotation interval [sec]. (default 60)
  -append
        append data to a file if it exists. (default true)
  -c string
        config file (other arguments will be ignored).
  -f string
        BPF rules.
  -i string
        device name (e.g. en0, eth0). (default "any")
  -offset int
        [deprecated] rotation interval offset [sec].
  -p    do NOT put into promiscuous mode. (default true)
  -s uint
        snapshot length. (default 65535)
  -sampling float
        sampling rate (0 <= p <= 1). (default 1)
  -t uint
        timeout of reading packets from interface [milli-sec]. (default 100)
  -utcoffset duration
        rotation interval offset from UTC [sec]. The negative value is also available.
  -v    show version and exit.
  -w string
        format of output file. (default "dump/%Y%m%d/traffic-%Y%m%d%H%M%S.pcap")
  -z string
        timezone used for output file. (default "UTC")
```


### Examples

Example-1: Capture traffic of HTTP (80/tcp) on the interface 'en0' and files are rotated at every 00:00 (UTC).

```sh
$ ./rcap -i en0 -w http-%Y%m%d.pcap -f "tcp and port 80" -T 86400
```

Example-2: Capture the whole traffic on the interface 'en0' and files are rotated at every 09:00 (i.e., 00:00 (JST)).
Plus, the filename of pcap files are filled with time (timezone "Asia/Tokyo").

```sh
$ ./rcap -i en0 -w traffic-%Y%m%d.pcap -T 86400 -utcoffset 9h -z Asia/Tokyo
```

You can find your timezone string here:
https://en.wikipedia.org/wiki/List_of_tz_database_time_zones

Example-3: Use configuration file.

```sh
# Copy and edit a configuration file.
$ cp -vi rcap.toml.orig rcap.toml
$ vim rcap.toml
... (edit) ...

$ ./rcap -c rcap.toml
```


### Configuration file

See [rcap.toml.orig](rcap.toml.orig).


### Systemd

[rcap.service.orig](rcap.service.orig) is a template unit file of systemd.
Edit it and start/enable the service.

```sh
# Copy systemd's unit file to systemd's directory.
$ cp -vi rcap.service.orig /etc/systemd/system/rcap.service

# Edit the unit file.
$ vim /etc/systemd/system/rcap.service
...(edit)...

# Start the rcap service.
$ systemctl start rcap

# (Optional) Enable the rcap service at startup.
$ systemctl enable rcap
```


## Alternatives

- tcpdump
- tshark
- dumpcap


## License

MIT License ([link](https://opensource.org/licenses/MIT)).


## Contact

mkt (E-mail: md.irohas at gmail.com)



CHANGELOG.md
# CHANGELOG

## v0.2

### v0.2.0 / 2023-07-05

- refactor: refactor codes
- feat: add configuration checker
- feat: add configuration reloader
- feat: print configuration to log
- feat: add fileAppend option to append packet data to the existing file
- feat: add utcOffset option for better offset value


## v0.1

### v0.1.0 / 2019-02-16

- feat: dumping packets as PCAP file
- feat: file rotation
- feat: flexible filename format (timezone-aware)
- feat: random sampling
- feat: configuration file
- feat: logging


rcap/run_test.go
package rcap

import (
	"errors"
	"path/filepath"
	"testing"
	"time"

	"github.com/google/gopacket"
)

func TestNewRunner(t *testing.T) {
	c := makeConfig()
	_, err := NewRunner(c)
	if err != nil {
		t.Errorf("nil is expected, but got '%v'.", err)
	}
}

func TestRunnerReloadWithFailure(t *testing.T) {
	c := makeConfig()
	r, _ := NewRunner(c)

	// empty config file
	if err := r.Reload(); err == nil {
		t.Error("err is expected, but got 'nil'.")
	}

	// invalid config
	c.Filename = "testdata/rcap-invalid-config.toml"
	if err := r.Reload(); err == nil {
		t.Error("err is expected, but got 'nil'.")
	}
}

func TestRunnerReload(t *testing.T) {
	c := makeConfig()
	r, _ := NewRunner(c)

	// valid config
	c.Filename = "testdata/rcap-good.toml"
	if err := r.Reload(); err != nil {
		t.Errorf("nil is expected, but got '%v'.", err)
	}

	// check if config values are updated.
	if c.Rcap.BpfRules == r.config.Rcap.BpfRules {
		t.Error("config is not updated.")
	}
}

func TestRunnerSetupReaderAndWriter(t *testing.T) {
	c := makeConfig()
	c.Rcap.FileFmt = filepath.Join(t.TempDir(), "traffic-%Y%m%d-%H%M%S.pcap")
	c.CheckAndFormat()
	r, _ := NewRunner(c)

	// NewReader fails.
	if err := r.setupReaderAndWriter(); err == nil {
		t.Error("err is expected, but got 'nil'.")
	}

	// NewReader and NewWriter succeed (reader is a dummy).
	r.reader, _ = openAndSetUpReader(c, "testdata/sample.pcap")
	if err := r.setupReaderAndWriter(); err != nil {
		t.Error("err is expected, but got 'nil'.")
	}
}

func TestRunnerGetTimestamp(t *testing.T) {
	c := makeConfig()
	c.CheckAndFormat()
	r, _ := NewRunner(c)

	metadata := gopacket.CaptureInfo{
		Timestamp: time.Unix(86400, 0),
	}
	pkterr := errors.New("dummy error")

	// Timestamp is from system time because of UseSystemTime.
	c.Rcap.UseSystemTime = true
	if ts := r.getTimestamp(metadata, nil); ts < 1688137200 {
		t.Error("'ts' is not from system time (UseSystemTime=true).")
	}

	// Timestamp is from system time because of pkterr.
	c.Rcap.UseSystemTime = false
	if ts := r.getTimestamp(metadata, pkterr); ts < 1688137200 {
		t.Error("'ts' is not from system time (UseSystemTime=false, pkterr!=nil).")
	}

	// Timestamp is from metadata.
	if ts := r.getTimestamp(metadata, nil); ts != 86400 {
		t.Errorf("'ts' is expected to be '%v', but got '%v'.", 86400, ts)
	}
}

func TestRunnnerPrintSamplingResult(t *testing.T) {
	c := makeConfig()
	c.CheckAndFormat()
	r, _ := NewRunner(c)

	// numCapturePackets == 0
	r.printSamplingResult()

	// numCapturePackets == 2, numSampledPackets == 1
	r.numCapturedPackets = 2
	r.numSampledPackets = 1
	r.printSamplingResult()
}

func TestRunnerDoSamplingAll(t *testing.T) {
	c := makeConfig()
	c.Rcap.Sampling = 1.0
	c.CheckAndFormat()

	r, _ := NewRunner(c)
	r.numStatsPackets = 10

	count := 0
	for i := 0; i < 100; i++ {
		if r.doSampling() {
			count++
		}
	}

	if count != 100 {
		t.Errorf("The probability of doSample seems to be invalid: (%v != 100)", count)
	}
}

func TestRunnerDoSampling10Percent(t *testing.T) {
	c := makeConfig()
	c.Rcap.Sampling = 0.1
	c.CheckAndFormat()

	r, _ := NewRunner(c)
	r.numStatsPackets = 10

	count := 0
	for i := 0; i < 100; i++ {
		if r.doSampling() {
			count++
		}
	}

	if !(count < 30) {
		t.Errorf("The probability of doSample seems to be invalid: (%v > 30)", count)
	}
}

func TestRunnerRunWithFailure(t *testing.T) {
	tempDir := t.TempDir()
	c := makeConfig()
	c.Rcap.FileFmt = filepath.Join(tempDir, "traffic-%Y%m%d-%H%M%S.pcap")
	c.CheckAndFormat()
	r, _ := NewRunner(c)

	// setupReaderAndWriter fails because of Device = 'any'.
	if err := r.Run(); err == nil {
		t.Error("err is expected, but got 'nil'.")
	}
}

func TestRunnerRun(t *testing.T) {
	tempDir := t.TempDir()
	c := makeConfig()
	c.Rcap.FileFmt = filepath.Join(tempDir, "traffic-%Y%m%d-%H%M%S.pcap")
	c.CheckAndFormat()
	r, _ := NewRunner(c)
	r.reader, _ = openAndSetUpReader(c, "testdata/sample.pcap")

	// EOF error.
	if err := r.Run(); err == nil {
		t.Error("err is expected, but got 'nil'.")
	}
}

func TestRunnerClose(t *testing.T) {
	// Setup reader and writer to be closed.
	tempDir := t.TempDir()
	c := makeConfig()
	c.Rcap.FileFmt = filepath.Join(tempDir, "traffic-%Y%m%d-%H%M%S.pcap")
	c.CheckAndFormat()
	r, _ := NewRunner(c)
	r.reader, _ = openAndSetUpReader(c, "testdata/sample.pcap")
	r.writer, _ = NewWriter(c, r.reader.LinkType())
	r.writer.openWriter(0)

	r.Close()
}

func TestRun(t *testing.T) {
	tempDir := t.TempDir()
	c := makeConfig()
	c.Rcap.FileFmt = filepath.Join(tempDir, "traffic-%Y%m%d-%H%M%S.pcap")
	c.CheckAndFormat()

	if err := Run(c); err == nil {
		t.Errorf("err is expected, but got 'nil'.")
	}
}


rcap/util_test.go
package rcap

import (
	"testing"
)

func TestRandom(t *testing.T) {
	// The Random function is just a wrapper of math.rand packet,
	// so this test case only checks its return value once.
	value := Random()
	if value < 0 || 1 <= value {
		t.Errorf("A number in [0.0, 1.0) is expected, but got %v.", value)
	}
}

func TestFileExists(t *testing.T) {
	if FileExists("testdata/rcap-good.toml") != true {
		t.Error("'testdata/rcap-good.toml' must be found, but not found.")
	}
	if FileExists("file-not-found") != false {
		t.Error("'file-not-found' must be not found, but found.")
	}
	if FileExists("") != false {
		t.Error("'' must be not found, but found.")
	}
}


rcap/config_test.go
package rcap

import (
	"testing"

	"github.com/go-playground/validator/v10"
	"github.com/google/go-cmp/cmp"
	"github.com/pelletier/go-toml"
)

func TestIsValidDevice(t *testing.T) {
	cases := []struct {
		// in
		device string
		// out
		isValid bool
	}{
		// device name
		{"any", true},
		{"lo", true},
		{"not-found-deviec", false},
	}

	for _, c := range cases {
		if isValidDevice(c.device) != c.isValid {
			t.Errorf("'%v' is expected, but got '%v' (device='%v').", c.isValid, !c.isValid, c.device)
		}
	}
}

func TestCheckDeviceAndBpf(t *testing.T) {
	cases := []struct {
		// in
		device string
		bpf    string
		// out
		isValid bool
	}{
		// device name
		{"any", "", true},
		{"lo", "", true},
		{"not-found-deviec", "", false},
		// bpf
		{"any", "ip", true},
		{"any", "ip and ((dst net 192.168.0.0/16 and dst port 53) or (src net 192.168.0.0/16 and src port 53))", true},
		{"any", "ip and (dst net 192.168.0.0/16 and dst port 53", false}, // lacks ) at the end of the expression
		{"any", "vlan and ip6", true}, // this is valid for ethernet but invalid for null
		{"any", "invalid bpf", false},
	}

	for _, c := range cases {
		err := CheckDeviceAndBpf(c.device, c.bpf, 65535)
		if c.isValid {
			if err != nil {
				t.Errorf("nil is expected, but got '%v': device='%v' bpf='%v'", err, c.device, c.bpf)
			}
		} else {
			if err == nil {
				t.Errorf("err is expected, but got 'nil': device='%v' bpf='%v'", c.device, c.bpf)
			}
		}
	}
}

func TestPrintToLog(t *testing.T) {
	config := &Config{Filename: "test"}
	config.PrintToLog()
}

// make config with default values.
func makeConfig() *Config {
	config := &Config{Filename: ""}
	toml.Unmarshal([]byte(`[rcap]`), config)
	return config
}

func TestConfigDefaultValues(t *testing.T) {
	got := makeConfig()

	expected := &Config{
		Filename: "",
		Rcap: RcapConfig{
			Device:        "any",
			SnapLen:       65535,
			Promisc:       true,
			ToMs:          100,
			BpfRules:      "",
			FileFmt:       "dump/%Y%m%d/traffic-%Y%m%d%H%M00.pcap",
			FileAppend:    true,
			Timezone:      "UTC",
			Location:      nil, // not set yet
			Interval:      60,
			Offset:        0,
			Sampling:      1.0,
			SamplingMode:  false, // not set yet (default)
			LogFile:       "",
			UseSystemTime: false,
		},
	}

	if !cmp.Equal(got, expected) {
		t.Errorf("'%#v' is expected, but got '%#v'.", got, expected)
		// diff := cmp.Diff(got, expected)
		// t.Errorf("diff: %v", diff)
	}
}

func TestConfigCheckAndFormat(t *testing.T) {
	c := makeConfig()

	if err := c.CheckAndFormat(); err != nil {
		t.Errorf("The default values are expected to be valid, but invalid: %v", err)
	}

	// check if Location is set.
	if c.Rcap.Location == nil {
		t.Errorf("The 'Location' value is still nil")
	}
}

func TestConfigCheckAndFormatWithFailure(t *testing.T) {
	c := makeConfig()
	r := &c.Rcap

	// invalid config
	r.Device = "" // required
	// r.SnapLen = -1	// SnapLen > 0, SnapLen is uint.
	r.ToMs = 1000                   // 0 < ToMs < 500
	r.Timezone = "invalid-timezone" // not-timezone string
	r.Interval = -1                 // Interval >= 0
	r.Offset = -1                   // Offset >= 0
	r.Sampling = 10.0               // 0.0 <= Sampling <= 1.0
	r.LogFile = "/some/dir/"        // empty or filepath

	err := c.CheckAndFormat()
	valErrs, _ := err.(validator.ValidationErrors)

	if len(valErrs) != 7 {
		t.Errorf("%d errors are expected, but %d errors are got.", 6, len(valErrs))
	}

	// for _, valErr := range valErrs {
	// 	t.Logf("validation error: %#v", valErr)
	// }

	c = makeConfig()
	r = &c.Rcap

	// invalid device and bpf
	r.Device = "not-found-device"
	r.BpfRules = "invalid bpf"

	err = c.CheckAndFormat()
	if err == nil {
		t.Error("err is expected, but got nil.")
	}
}

func TestLoadConfig(t *testing.T) {
	if _, err := LoadConfig("testdata/rcap-good.toml"); err != nil {
		t.Errorf("nil is expected, but got '%v'.", err)
	}
}

func TestLoadConfigWithFailure(t *testing.T) {
	cases := []struct {
		// in
		filename string
		// out
		errMsg string
	}{
		{"", "open : no such file or directory"},
		{"file-not-found", "open file-not-found: no such file or directory"},
		{"testdata/rcap-invalid-toml.toml", "(2, 1): parsing error: keys cannot contain { character"},
		{"testdata/rcap-invalid-config.toml", "invalid config values: Key: 'Config.Rcap.Offset' Error:Field validation for 'Offset' failed on the 'gte' tag"},
	}

	for _, c := range cases {
		t.Logf("test with '%v'", c.filename)

		if _, err := LoadConfig(c.filename); err.Error() != c.errMsg {
			t.Errorf("'%v' is expected, but got '%v'.", c.errMsg, err.Error())
		}
	}
}


main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"os"

	"github.com/md-irohas/rcap-go/rcap"
)

var (
	Version = "(unset)" // Version
)

func main() {
	var configFile string
	var showVersion bool
	var fileConfig, argsConfig, config *rcap.Config
	var err error

	// Meta flags.
	flag.StringVar(&configFile, "c", "", "config file (other arguments will be ignored).")
	flag.BoolVar(&showVersion, "v", false, "show version and exit.")

	// rcap config from command line.
	argsConfig = &rcap.Config{}
	r := &argsConfig.Rcap

	// rcap config flags.
	flag.StringVar(&r.Device, "i", "any", "device name (e.g. en0, eth0).")
	flag.UintVar(&r.SnapLen, "s", 65535, "snapshot length.")
	flag.BoolVar(&r.Promisc, "p", true, "do NOT put into promiscuous mode.")
	flag.UintVar(&r.ToMs, "t", 100, "timeout of reading packets from interface [milli-sec].")
	flag.StringVar(&r.BpfRules, "f", "", "BPF rules.")
	flag.StringVar(&r.FileFmt, "w", "dump/%Y%m%d/traffic-%Y%m%d%H%M%S.pcap", "format of output file.")
	flag.BoolVar(&r.FileAppend, "append", true, "append data to a file if it exists. to disable, add -append=false as argument.")
	flag.StringVar(&r.Timezone, "z", "UTC", "timezone used for output file.")
	flag.Int64Var(&r.Interval, "T", 60, "rotation interval [sec]. to disable rotation, set 0.")
	flag.Int64Var(&r.Offset, "offset", 0, "[deprecated] rotation interval offset [sec]. use -utcoffset instead.")
	flag.DurationVar(&r.UTCOffset, "utcoffset", 0, "rotation interval offset from UTC. The negative value is also available. see https://pkg.go.dev/time#Duration for the format.")
	flag.Float64Var(&r.Sampling, "sampling", 1.0, "sampling rate (0.0 <= p <= 1.0).")
	flag.StringVar(&r.LogFile, "L", "", "[deprecated] log file.")
	flag.BoolVar(&r.UseSystemTime, "S", false, "use system time as a time source of rotation (default: use packet-captured time).")
	flag.Parse()

	if showVersion {
		fmt.Println(Version)
		os.Exit(0)
	}

	log.Printf("rcap version: %v", Version)

	if configFile != "" {
		log.Printf("load config: %v", configFile)

		// Load config from file and check its parameters.
		fileConfig, err = rcap.LoadConfig(configFile)
		if err != nil {
			log.Fatalf("failed to load config from file: %v", err)
		}

		config = fileConfig
	} else {
		log.Println("load config from command-line.")

		// Check config parsed from command-line.
		err = argsConfig.CheckAndFormat()
		if err != nil {
			log.Fatalf("failed to load config from command-line: %v", err)
		}

		config = argsConfig
	}

	config.PrintToLog()

	if err := rcap.Run(config); err != nil {
		log.Fatalf("fatal error: %v", err)
	}
}


rcap.go
package main

import (
	"flag"
	"fmt"
	//	"github.com/google/gopacket/layers"
	"github.com/google/gopacket/pcap"
	"github.com/google/gopacket/pcapgo"
	"github.com/jehiah/go-strftime"
	"github.com/pelletier/go-toml"
	"log"
	"math/rand"
	"os"
	"os/signal"
	"path/filepath"
	"strconv"
	"syscall"
	"time"
	// debug
	// "reflect"
)

const (
	// SamplingDump holds the number of packets to dump sampling results.
	SamplingDump = 10000
)

var version = "0.0.6"

var (
	// Params used for libpcap.
	device   string
	snaplen  uint
	promisc  bool
	toMs     uint
	bpfRules string

	// Params used for this program.
	fileFmt       string  // format of output file
	timezone      string  // timezone used for fileFmt
	interval      int64   // rotation interval [sec]
	offset        int64   // rotation interval offset [sec]
	sampling      float64 // sampling rate (probability, from 0 to 1)
	logFile       string  // path to log file
	useSystemTime bool    // use system time as a time source of rotation
)

func fileExists(filename string) bool {
	_, err := os.Stat(filename)
	return err == nil
}

func main() {
	var f *os.File
	var w *pcapgo.Writer
	var showVersion bool
	var cnfFile string

	// Parse params from command-line arguments.
	flag.StringVar(&device, "i", "", "device name (e.g. en0, eth0).")
	flag.UintVar(&snaplen, "s", 65535, "snapshot length.")
	flag.BoolVar(&promisc, "p", true, "do NOT put into promiscuous mode.")
	flag.UintVar(&toMs, "t", 100, "timeout of reading packets from interface [milli-sec].")
	flag.StringVar(&bpfRules, "f", "", "BPF rules.")
	flag.StringVar(&fileFmt, "w", "pcap/%Y%m%d/%Y%m%d-%H%M%S.pcap", "format of output file.")
	flag.StringVar(&timezone, "z", "Local", "timezone used for output file.")
	flag.Int64Var(&interval, "T", 60, "rotation interval [sec].")
	flag.Int64Var(&offset, "offset", 0, "rotation interval offset [sec].")
	flag.Float64Var(&sampling, "sampling", 1, "sampling rate (0 <= p <= 1).")
	flag.StringVar(&logFile, "L", "", "log file.")
	flag.StringVar(&cnfFile, "c", "", "config file (other arguments are ignored).")
	flag.BoolVar(&showVersion, "v", false, "show version and exit.")
	flag.BoolVar(&useSystemTime, "S", false, "use system time as a time source of rotation (default: use packet-captured time).")
	flag.Parse()

	if showVersion {
		fmt.Println("version:", version)
		return
	}

	// Parse params from config file. Params in the command-line arguments are ignored.
	if cnfFile != "" {
		cnf, err := toml.LoadFile(cnfFile)
		if err != nil {
			log.Fatal(err)
		}

		device = cnf.Get("rcap.device").(string)
		snaplen = uint(cnf.Get("rcap.snaplen").(int64))
		promisc = cnf.Get("rcap.promisc").(bool)
		toMs = uint(cnf.Get("rcap.toMs").(int64))
		bpfRules = cnf.Get("rcap.bpfRules").(string)
		fileFmt = cnf.Get("rcap.fileFmt").(string)
		timezone = cnf.Get("rcap.timezone").(string)
		interval = cnf.Get("rcap.interval").(int64)
		offset = cnf.Get("rcap.offset").(int64)
		sampling = cnf.Get("rcap.sampling").(float64)
		logFile = cnf.Get("rcap.logFile").(string)
		useSystemTime = cnf.Get("rcap.useSystemTime").(bool)
	}

	// Init logging.
	log.SetFlags(log.Ldate | log.Ltime)

	if logFile != "" {
		logger, err := os.OpenFile(logFile, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0640)
		if err != nil {
			log.Fatal(err)
		}

		log.SetOutput(logger)
		log.Println("open a log file:", logFile)
	}

	// Init signal handling.
	sigc := make(chan os.Signal, 1)
	signal.Notify(sigc, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM)

	go func() {
		s := <-sigc
		log.Println("SIGNAL:", s)
		f.Close()
		os.Exit(0)
	}()

	// Init random generator.
	r := rand.New(rand.NewSource(time.Now().UnixNano()))

	// Lookup a default device if no device name is given.
	if device == "" {
		log.Println("no device is set. look up devices.")

		devices, err := pcap.FindAllDevs()
		if err != nil {
			log.Fatal(err)
		}

		for _, dev := range devices {
			device = dev.Name
			break
		}

		log.Println("find device:", device)
	}

	// Open a device and set a BPF filter.
	handle, err := pcap.OpenLive(device, int32(snaplen), promisc, time.Duration(toMs)*time.Millisecond)
	if err != nil {
		log.Fatal(err)
	}
	defer handle.Close()

	log.Println("open device:", device)

	linkType := handle.LinkType()
	log.Println("linktype:", linkType)

	if bpfRules != "" {
		err := handle.SetBPFFilter(bpfRules)
		if err != nil {
			log.Fatal(err)
		}

		log.Println("set filter:", bpfRules)
	}

	// Load location from timezone.
	loc, err := time.LoadLocation(timezone)
	if err != nil {
		log.Fatal(err)
	}

	log.Println("set timezone to", timezone)

	// Set sampling mode.
	samplingMode := (sampling < 1)

	log.Println("set sampling rate:", sampling)

	if useSystemTime {
		log.Println("set time source: system")
	} else {
		log.Println("set time source: packet")
	}

	var lstTime int64 = 0
	var numPackets, numCapPackets uint64 = 0, 0

	// Read packets from device and dump them into a file.
	for {
		data, capinfo, pkterr := handle.ZeroCopyReadPacketData()

		var curTime int64

		if useSystemTime {
			curTime = time.Now().Unix()
		} else {
			if pkterr != nil {
				curTime = time.Now().Unix()
			} else {
				curTime = capinfo.Timestamp.Unix()
			}
		}

		if f != nil {
			// If file need to be rotated, close the file and reset the file variable.
			if (curTime > lstTime) && (interval > 0) && (curTime%interval == offset) {
				f.Close()
				f = nil
			}
		}

		if f == nil {
			lstTime = curTime

			// Convert unix time to native time.
			tmpTime := time.Unix(lstTime, 0)
			tmpTime = tmpTime.In(loc)

			// Fill format of date and time in fileFmt.
			fileName := strftime.Format(fileFmt, tmpTime)

			// If the filename already exists, find alternative filenames.
			// e.g. foobar.pcap -> foobar-1.pcap
			for i := 1; fileExists(fileName); i++ {
				log.Println("file already exists:", fileName)

				fileFmtExt := filepath.Ext(fileFmt)
				fileFmtBase := fileFmt[0 : len(fileFmt)-len(fileFmtExt)]
				newFileFmt := fileFmtBase + "-" + strconv.Itoa(i) + fileFmtExt
				fileName = strftime.Format(newFileFmt, tmpTime)
			}

			// Make a directory for output files.
			dirName := filepath.Dir(fileName)
			err := os.MkdirAll(dirName, 0755)
			if err != nil {
				log.Fatal(err)
			}

			// Make a new file and write a PCAP header.
			f, err = os.Create(fileName)
			if err != nil {
				log.Fatal(err)
			}

			w = pcapgo.NewWriter(f)
			w.WriteFileHeader(uint32(snaplen), linkType)

			log.Printf("capture %v packets.\n", numCapPackets)
			log.Println("dump packets into a file:", fileName)

			numPackets = 0
			numCapPackets = 0
		}

		if pkterr != nil {
			// Do NOT log messages when it is timeouted.
			if pkterr != pcap.NextErrorTimeoutExpired {
				log.Println(pkterr)
			}
			continue
		}

		// Sample packets when sampling mode.
		if samplingMode {
			sample := (r.Float64() <= sampling)

			numPackets++
			if sample {
				numCapPackets++
			}

			if numPackets%SamplingDump == 0 {
				log.Printf("sampling result: %d/%d\n", numCapPackets, numPackets)
			}

			if !sample {
				continue
			}
		} else {
			numPackets++
			numCapPackets++
		}

		if f != nil {
			err := w.WritePacket(capinfo, data)
			if err != nil {
				log.Println(err)
			}
		}
	}
}


rcap/config.go
package rcap

import (
	"fmt"
	"log"
	"os"
	"time"

	"github.com/go-playground/validator/v10"
	"github.com/google/gopacket/layers"
	"github.com/google/gopacket/pcap"
	"github.com/pelletier/go-toml"
)

// Config struct is a root section of rcap-go configuration.
type Config struct {
	// Filename which has this configuration.
	Filename string

	// Rcap struct is a main section of rcap-go configuration.
	Rcap RcapConfig `toml:"rcap"`
}

// RcapConfig struct is a main section of rcap-go configuration.
// TODO: Add validator of Device and BpfRules.
type RcapConfig struct {
	// Params for libpcap.
	Device   string `toml:"device" default:"any" validate:"required"`    // Device name.
	SnapLen  uint   `toml:"snaplen" default:"65535" validate:"gte=0"`    // Snap length.
	Promisc  bool   `toml:"promisc" default:"true"`                      // Promiscuous mode.
	ToMs     uint   `toml:"toMs" default:"100" validate:"gte=1,lte=500"` // Timeout when no packets are captured.
	BpfRules string `toml:"bpfRules" default:""`                         // BPF rules.

	// Params for this program.
	FileFmt       string         `toml:"fileFmt" default:"dump/%Y%m%d/traffic-%Y%m%d%H%M00.pcap" validate:"filepath"` // Path to PCAP files.
	FileAppend    bool           `toml:"fileAppend" default:"true"`                                                   // Append data if the file exists.
	Timezone      string         `toml:"timezone" default:"UTC" validate:"timezone"`                                  // Timezone used for FileFmt.
	Location      *time.Location // Location data (i.e., Timezone)
	Interval      int64          `toml:"interval" default:"60" validate:"gte=0"`        // Rotation interval (in second).
	Offset        int64          `toml:"offset" default:"0" validate:"gte=0"`           // Deprecated: Rotation offset (in second).
	UTCOffset     time.Duration  `toml:"utcOffset" default:"0"`                         // Rotation offset from UTC (in second).
	Sampling      float64        `toml:"sampling" default:"1.0" validate:"gte=0,lte=1"` // Sampling rate.
	SamplingMode  bool           // Sampling mode.
	LogFile       string         `toml:"logFile" default:"" validate:"omitempty,filepath"` // Deprecated: Log file.
	UseSystemTime bool           `toml:"useSystemTime" default:"false"`                    // Use system time or packet-captured time.
}

func isValidDevice(name string) bool {
	devices, err := pcap.FindAllDevs()
	if err != nil {
		panic(err)
	}

	for _, device := range devices {
		if device.Name == name {
			return true
		}
	}

	return false
}

func isValidBpf(bpf string, device string, captureLength uint) bool {
	h, err := pcap.OpenLive(device, int32(captureLength), false, pcap.BlockForever)
	if err == nil {
		defer h.Close()
		return h.SetBPFFilter(bpf) == nil
	}

	// For test
	linkType := layers.LinkTypeEthernet
	log.Printf("The device linktype could not be detected. '%v' is used anyway.", linkType)
	_, err = pcap.CompileBPFFilter(linkType, int(captureLength), bpf)

	return err == nil
}

func CheckDeviceAndBpf(device string, bpf string, captureLength uint) error {
	if !isValidDevice(device) {
		return fmt.Errorf("invalid device: '%v'", device)
	}
	if !isValidBpf(bpf, device, captureLength) {
		return fmt.Errorf("invalid BPF: '%v'", bpf)
	}
	return nil
}

// CheckAndFormat method checks and formats the values in the configuration,
// and returns an error if the configuration is invalid.
func (c *Config) CheckAndFormat() error {
	validate := validator.New()

	if err := validate.Struct(c); err != nil {
		return err
	}
	if err := CheckDeviceAndBpf(c.Rcap.Device, c.Rcap.BpfRules, c.Rcap.SnapLen); err != nil {
		return err
	}

	// no error is returned from LoadLocation because validator checks timezone value.
	c.Rcap.Location, _ = time.LoadLocation(c.Rcap.Timezone)
	c.Rcap.SamplingMode = (c.Rcap.Sampling < 1.0)

	return nil
}

// PrintToLog method prints config values to its log.
func (c *Config) PrintToLog() {
	r := &c.Rcap

	log.Printf("==== RCAP Config ====\n")
	log.Printf("- Filename: %v\n", c.Filename)
	log.Printf("- RCAP:\n")
	log.Printf("  - device:	%v\n", r.Device)
	log.Printf("  - snaplen:	%v\n", r.SnapLen)
	log.Printf("  - promisc:	%v\n", r.Promisc)
	log.Printf("  - toMs:	%v\n", r.ToMs)
	log.Printf("  - bpfRules:	%v\n", r.BpfRules)
	log.Printf("  - fileFmt:	%v\n", r.FileFmt)
	log.Printf("  - fileAppend:	%v\n", r.FileAppend)
	log.Printf("  - timezone:	%v (location: %v)\n", r.Timezone, r.Location)
	log.Printf("  - interval:	%v\n", r.Interval)
	log.Printf("  - offset:	%v\n", r.Offset)
	log.Printf("  - utcOffset:	%v\n", r.UTCOffset)
	log.Printf("  - sampling:	%v (samplingMode: %v)\n", r.Sampling, r.SamplingMode)
	log.Printf("  - useSystemTime:	%v\n", r.UseSystemTime)
	log.Printf("=====================\n")
}

// LoadConfig loads a configuration from the given filename and returns an
// instance of Config struct.
func LoadConfig(filename string) (*Config, error) {
	// empty filename and file-not-found error are handled here.
	str, err := os.ReadFile(filename)
	if err != nil {
		return nil, err
	}

	config := &Config{Filename: filename}
	if err := toml.Unmarshal(str, config); err != nil {
		return nil, err
	}

	if err := config.CheckAndFormat(); err != nil {
		valErrs, ok := err.(validator.ValidationErrors)
		if ok {
			for _, valErr := range valErrs {
				log.Println(valErr.Error())
			}
		}
		return nil, fmt.Errorf("invalid config values: %w", err)
	}

	return config, nil
}
